internal/tracer.go

```diff
@@ -119,17 +128,56 @@ func (t *tracingContextPropagator) InjectFromWorkflow(
 	if spanContext == nil {
 		return nil
 	}
-	return t.tracer.Inject(spanContext, opentracing.TextMap, tracingWriter{hw})
+
+	return t.writeSpanContextToHeader(spanContext, hw)
 }
 
 func (t *tracingContextPropagator) ExtractToWorkflow(
 	ctx Context,
 	hr HeaderReader,
 ) (Context, error) {
-	spanContext, err := t.tracer.Extract(opentracing.TextMap, tracingReader{hr})
+	spanContext, err := t.readSpanContextFromHeader(hr)
 	if err != nil {
-		// did not find a tracing span, just return the current context
 		return ctx, nil
 	}
 	return contextWithSpan(ctx, spanContext), nil
 }
+
+func (t *tracingContextPropagator) writeSpanContextToHeader(spanContext opentracing.SpanContext, hw HeaderWriter) error {
+	tracerData := make(map[string]string)
+
+	err := t.tracer.Inject(spanContext, opentracing.TextMap, tracingWriter{tracerData})
+	if err != nil {
+		return err
+	}
+
+	tracerPayload, err := converter.GetDefaultDataConverter().ToPayload(tracerData)
+	if err != nil {
+		return err
+	}
+
+	hw.Set(tracerHeaderKey, tracerPayload)
+	return nil
+}
+
+func (t *tracingContextPropagator) readSpanContextFromHeader(hr HeaderReader) (opentracing.SpanContext, error) {
```

Comment:
This method should take tracerData map as a parameter and deserialization should happen in the caller